### PR TITLE
general cleanup and maintenance

### PR DESCRIPTION
**Pull request by AnthonyMichaelTDM**
- changed `StatusCode` from `i31` to `u16`.
- refactored logic
- expanded documentation
- added TODOs
- made some functions more generic (to improve API w/o causing breaking changes)

this is just a start, there's a lot of TODOs I still need to tackle as well

### CODE REVIEW

**Comment by Wulf** [comment]
Whao! Thanks for taking the time to do this. Just to make it easier to review and stuff, could I ask you to split changes into smaller bites and make separate merge requests for all of them?

It's easier to revert and bisect for bugs that way too :)

**Comment by AnthonyMichaelTDM** [comment]
sure, I can have this as base then do subsequent PR's to address the TOOD's and whatnot

**Comment by AnthonyMichaelTDM** [comment]
I forgot why I hadn't merged this yet, it's been a while...
I plan on looking over it again soon though

**Comment by AnthonyMichaelTDM** [comment]
Okay I think it's good now, gonna do one final review before I merge though

File: create-rust-app/src/auth/permissions/mod.rs [hunk 1]
```diff
@@ -45,6 +50,11 @@ impl Role {
     /// assigns every role in `roles` to the User whose id is [`user_id`](`ID`)
     ///
     /// returns true if successful
+    ///
+    /// # Errors
+    /// * infallible
+    ///
+    /// TODO: don't return a result if we never fail, or return a result and not a bool
     pub fn assign_many(db: &mut Connection, user_id: ID, roles: Vec<String>) -> Result<bool> {
```
Comment (AnthonyMichaelTDM) [inline create-rust-app/src/auth/permissions/mod.rs:58]: regarding these todo statements, I think that returning a `Result<()>` is more ergonomic for the user than `bool`, just because then the function signiture explicitly tells users that the returned value indicates success/failure rather than having them need to read the documentation to know that.

File: create-rust-app/src/auth/controller.rs [hunk 2]
```diff
@@ -15,17 +15,15 @@ lazy_static! {
     pub static ref ARGON_CONFIG: argon2::Config<'static> = argon2::Config {
         variant: argon2::Variant::Argon2id,
         version: argon2::Version::Version13,
-        secret: match std::env::var("SECRET_KEY") {
-            Ok(s) => Box::leak(s.into_boxed_str()).as_bytes(),
-            Err(_) => panic!("No SECRET_KEY environment variable set!"),
-        },
+        secret: std::env::var("SECRET_KEY").map_or_else(|_| panic!("No SECRET_KEY environment variable set!"), |s| Box::leak(s.into_boxed_str()).as_bytes()),
```
Comment (AnthonyMichaelTDM) [inline create-rust-app/src/auth/controller.rs:18]: we might be able to just use a `.map(...).expect(...)` since we panic anyway

File: create-rust-app/src/dev/backend_compiling_server.rs [hunk 3]
```diff
@@ -207,6 +210,8 @@ pub async fn start(
     println!("backend compilation server stopped.");
 }
 
+/// TODO: do we need to pass the `ws_s` by value here?
```
Comment (AnthonyMichaelTDM) [inline create-rust-app/src/dev/backend_compiling_server.rs:213]: answer: it doesn't matter, `Sender` is just a wrapper around an `Arc` anyway

File: create-rust-app/src/dev/controller.rs [hunk 4]
```diff
@@ -37,12 +43,20 @@ pub fn query_db(db: &Database, body: &MySqlQuery) -> Result<String, diesel::resu
...
 /// /db/is-connected
+///
+/// # Panics
+/// * cannot connect to the database
+#[must_use]
 pub fn is_connected(db: &Database) -> bool {
     let mut db = db.pool.clone().get().unwrap();
     let is_connected = sql_query("SELECT 1;").execute(&mut db);
     is_connected.is_err()
 }
```
Comment (AnthonyMichaelTDM) [inline create-rust-app/src/dev/controller.rs:54]: if we fail to connect to the db, then we obviously aren't connected and should return false instead of panicing... this wasn't introduced in this PR so I'll fix it in a new one, but still.

```suggestion
/// /db/is-connected
#[must_use]
pub fn is_connected(db: &Database) -> bool {
    let Ok(mut db) = db.pool.clone().get() else {
        return false;
    };
    let is_connected = sql_query("SELECT 1;").execute(&mut db);
    is_connected.is_err()
}
```

File: create-rust-app/src/dev/mod.rs [hunk 5]
```diff
@@ -300,7 +304,7 @@ fn check_exit(state: &DevState) {
 }
 
 async fn listen_for_signals(signal_tx: tokio::sync::broadcast::Sender<DevServerEvent>) {
-    let (ev_s, ev_r) = priority::bounded::<Event, Priority>(1024);
+    let (event_sender, event_receiver) = priority::bounded::<Event, Priority>(1024);
     let (er_s, mut er_r) = tokio::sync::mpsc::channel(64);
```
Comment (AnthonyMichaelTDM) [inline create-rust-app/src/dev/mod.rs:308]: should rename this to `error_sender` and `error_receiver` respectively, again, will do in a different PR

File: create-rust-app/src/mailer.rs [hunk 6]
```diff
@@ -165,6 +172,12 @@ impl Mailer {
     /// * `subject` - subject field of the email
     /// * `text` - text content of the email
     /// * `html` - html content of the email
+    ///
+    /// # Panics
+    ///
+    /// panis if the `to` argument is not a valid email address
+    ///
+    /// TODO: wouldn't it be better to instead require the `to` argument be some wrapper around a string that is always a valid email address?
```
Comment (AnthonyMichaelTDM) [inline create-rust-app/src/mailer.rs:180]: answer: yes it would, but that's something to do in a different PR

**Review by AnthonyMichaelTDM — COMMENTED** [review]
Looks good

File: create-rust-app/src/dev/backend_compiling_server.rs (hunk 3 above)
Comment (AnthonyMichaelTDM) [reply create-rust-app/src/dev/backend_compiling_server.rs:213]: resolved by #371

File: create-rust-app/src/auth/permissions/mod.rs (hunk 1 above)
Comment (AnthonyMichaelTDM) [reply create-rust-app/src/auth/permissions/mod.rs:58]: resolved by #372

File: create-rust-app/src/dev/mod.rs (hunk 5 above)
Comment (AnthonyMichaelTDM) [reply create-rust-app/src/dev/mod.rs:308]: fixed in #373

File: create-rust-app/src/dev/controller.rs (hunk 4 above)
Comment (AnthonyMichaelTDM) [reply create-rust-app/src/dev/controller.rs:54]: fixed in #373
